core/threaded/scheduler_GEDF_NP.c

```diff
@@ -223,91 +105,138 @@ void lf_sched_init(environment_t* env, size_t number_of_workers, sched_params_t*
 
+  // Iterate until the stop_tag is reached or the event queue is empty.
+  while (!scheduler->should_stop) {
+    reaction_t* reaction_to_return = (reaction_t*)pqueue_peek(scheduler->custom_data->reaction_q);
     if (reaction_to_return != NULL) {
-      // Got a reaction
-      return reaction_to_return;
+      // Found a reaction.  Check the level.  Notice that because of deadlines, the current level
+      // may advance to the maximum and then back down to 0.
+      if (LF_LEVEL(reaction_to_return->index) == scheduler->custom_data->current_level) {
+        // Found a reaction at the current level.
+        LF_PRINT_DEBUG("Scheduler: Worker %d found a reaction at level %zu.", worker_number,
+                       scheduler->custom_data->current_level);
+        // Remove the reaction from the queue.
+        pqueue_pop(scheduler->custom_data->reaction_q);
+
+        // If there is another reaction at the current level and an idle thread, then
+        // notify an idle thread.
+        reaction_t* next_reaction = (reaction_t*)pqueue_peek(scheduler->custom_data->reaction_q);
+        if (next_reaction != NULL && LF_LEVEL(next_reaction->index) == scheduler->custom_data->current_level &&
+            scheduler->number_of_idle_workers > 0) {
+          // Notify an idle thread.
+          LF_COND_SIGNAL(&scheduler->custom_data->reaction_q_changed);
+        }
+        LF_MUTEX_UNLOCK(&scheduler->env->mutex);
+        return reaction_to_return;
+      } else {
+        // Found a reaction at a level other than the current level.
+        LF_PRINT_DEBUG("Scheduler: Worker %d found a reaction at level %lld. Current level is %zu", worker_number,
+                       LF_LEVEL(reaction_to_return->index), scheduler->custom_data->current_level);
+        // We need to wait to advance to the next level or get a new reaction at the current level.
+        if (scheduler->number_of_idle_workers == scheduler->number_of_workers - 1) {
+          // All other workers are idle.  Advance to the next level.
+          if (++scheduler->custom_data->current_level > scheduler->max_reaction_level) {
+            // Since the reaction queue is not empty, we must be cycling back to level 0 due to deadlines
+            // having been given precedence over levels.  Reset the current level to 1.
+            scheduler->custom_data->current_level = 0;
+          }
+          LF_PRINT_DEBUG("Scheduler: Advancing to next reaction level %zu.", scheduler->custom_data->current_level);
+#ifdef FEDERATED
+          // In case there are blocking network input reactions at this level, stall.
+          lf_stall_advance_level_federation_locked(scheduler->custom_data->current_level);
+#endif
+        } else {
+          // Some workers are still working on reactions on the current level.
+          // Wait for them to finish.
+          wait_for_other_workers_to_finish(scheduler, worker_number);
+        }
+      }
```

Comment:
This could be factored out into a separate `static` procedure with a signature like `advance_level_or_wait(lf_scheduler_t* scheduler, int worker_number)`

Reply:
Again, I have opted for a somewhat smaller function `advance_level` to keep the decision logic in this main procedure.